filestack/utils/intelligent_ingestion.py

```diff
@@ -0,0 +1,435 @@
+
+    def _manage_upload_process(self):
+        self._feed_uploaders()
+
+        while self.parts:
+            response = self.response_q.get(block=True)
+            log.info('Got response %s', response)
+
+            if response['worker'] == 'uploader':
+                self.parts[response['part']]['currently_processed'] -= 1
+                self._currently_processed -= 1
+                old_chunk = response['chunk']
+
+                if not response['success']:
+                    log.warning('Failed response received %s', response)
+
+                    if response['delay']:
+                        # this means uploader got a response, but it wasn't ok (status code >= 400)
+                        # resubmit with requested delay if max delay not exceeded
+                        if response['delay'] > MAX_DELAY:
+                            log.error('Max delay exceeded for chunk %s', old_chunk)
+                            return
+                        self._submit_upload_job(response['part'], old_chunk, delay=response['delay'])
+                        continue
+
+                    if old_chunk['size'] <= self.chunk_size:
+                        log.info(
+                            'Failed to upload %s bytes. Changing chunk size from %s to %s bytes',
+                            old_chunk['size'], self.chunk_size, self.chunk_size / 2
+                        )
+                        self.chunk_size /= 2
+
+                        if self.chunk_size < 32 * 1024:
+                            log.error('Minimal chunk size failed')
+                            return
+
+                    new_chunks = self._split_chunk(old_chunk)
+                    for new_chunk in reversed(new_chunks):
+                        self.parts[response['part']]['chunks'].append(new_chunk)
+
+                    self._feed_uploaders()
+                    continue
+
+                if not self.parts[response['part']]['chunks'] and self.parts[response['part']]['currently_processed'] == 0:
+                    log.info('No more chunks for part %s, time to commit', response['part'])
+                    self.commit_q.put({
+                        'apikey': self.apikey,
+                        'uri': self.start_response['uri'],
+                        'region': self.start_response['region'],
+                        'upload_id': self.start_response['upload_id'],
+                        'size': self.filesize,
+                        'part': response['part'],
+                        'store_location': self.storage,
+                        'filename': self.filename,
+                    })
+
+                self._feed_uploaders()
+
+            elif response['worker'] == 'committer':
+                log.info('Got commit done message %s', response)
+                log.info('Removing part %s', response['part'])
+                self.parts.pop(response['part'])
+
+        if self._get_next_chunk()[1] is None:
```

Comment:
this means there are no more chunks.

`_get_next_chunk()` returns a (part_num, chunk) tuple. It there are no more chunks, it returns (None, None). I'm just checking for chunk part ;)